### PR TITLE
SAA-645 Require note on savings fund transfer form

### PR DESCRIPTION
**Pull request by jeff-clay**
## Description

- Makes the Note field required on one-time fund transfers (previously labeled "Note (Optional)", which produced a confusing SAA 422 when left blank).
- Defaults the Note to `<lastName> Savings Fund Transfer in MPDX` using the current user's `lastName`. Editable — clearing it triggers the required-field error.
- Front-end enforcement only; recurring transfers are unaffected (their description is server-synthesized).

**Related:** [SAA-645](https://jira.cru.org/browse/SAA-645)

## Testing

- Open **HR Tools → Savings Fund Transfer → New Transfer**, **One Time** schedule.
- Confirm Note is pre-filled with `<your-last-name> Savings Fund Transfer in MPDX`.
- Submit unchanged → succeeds.
- Open another new transfer, clear Note (or enter only whitespace), submit → "Note is required" inline error, no submission.
- Switch to **Monthly** schedule — no Note field shown, recurring submits normally.

## Checklist:

- [x] I have given my PR a title with the format "MPDX-(JIRA#) (summary sentence max 80 chars)" — using `SAA-645` since this work is tracked in the SAA project's Jira
- [x] I have applied the appropriate labels (_Add the label "Preview" to automatically create a preview environment_)
- [x] I have run the Claude Code `/pr-review` command locally and fixed any relevant suggestions
- [x] I have requested a review from another person on the project
- [ ] I have tested my changes in preview or in staging
- [x] I have cleaned up my commit history

### CODE REVIEW

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 1]
```diff
@@ -169,6 +179,10 @@ export const TransferModal: React.FC<TransferModalProps> = ({
   const type = data.type || TransferTypeEnum.New;
   const isNew = type === TransferTypeEnum.New;
 
+  const defaultNote = lastName
```
Comment (jeff-clay) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx]: [Nice-to-have] Consider trimming `lastName` before interpolating, so an upstream value like `'  Sleight  '` doesn't produce a default with leading/trailing whitespace. `const trimmed = lastName?.trim(); const defaultNote = trimmed ? t(...) : '';`

**Review by jeff-clay — COMMENTED** [review]
Small, well-scoped form/validation change. All SAA-645 acceptance criteria met (required field with visual asterisk, whitespace-only rejected via `.trim().min(1)`, mutation never fires when invalid). Default-note feature is a documented scope addition per product owner. Refactor to `useGetUserQuery` over parsing `staffAccount.name` is the right architectural call — Apollo's cache makes the duplicate hook call free. No must-fix issues; a few nice-to-have hardening suggestions and one optional whitespace test below.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx (hunk 1 above)
Comment (jeff-clay) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx]: [Nice-to-have] `defaultNote` only runs through Formik `initialValues` at mount. If `useGetUserQuery` is still pending on first open (cold Apollo cache), the modal opens with no default. `UserPreferenceProvider` warms the cache at app root so it's unlikely, but if you see it in QA, `enableReinitialize={true}` on Formik or a render-guard in TransfersPage are the two fixes.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 2]
```diff
@@ -135,18 +135,28 @@ const transferSchema = (locale: string) =>
       .number()
       .required(i18n.t('Amount is required'))
       .min(0.01, i18n.t('Amount must be at least $0.01')),
-    note: yup.string().nullable(),
+    note: yup.string().when('schedule', {
```
Comment (jeff-clay) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx:139]: [Mentoring] `.trim().min(1).required()` is the canonical Yup pattern for non-blank — `.required()` alone would still let `'   '` through, because Yup treats any non-undefined string as 'present'. The trim transformation strips whitespace, then min(1) catches the empty result.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 3]
```diff
@@ -256,7 +270,7 @@ export const TransferModal: React.FC<TransferModalProps> = ({
           status: data.transfer.status ?? '',
           transferDate: data.transfer.transferDate ?? getToday(),
           endDate: data.transfer.endDate ?? null,
-          note: data.transfer.note ?? '',
+          note: data.transfer.note || defaultNote,
```
Comment (jeff-clay) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx:277]: [Mentoring] `||` vs `??` matters here. With `||`, empty strings fall through to `defaultNote` (what we want, since `transferDefaultData.note = ''` for new transfers). With `??`, only nullish values fall through and empty stays. The original `?? ''` was just normalizing nullish to empty; the new `|| defaultNote` is an actual upgrade-to-default.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx [hunk 4]
```diff
@@ -167,6 +170,34 @@ describe('TransferModal', () => {
       ).toBeInTheDocument();
     });
 
+    it('should require a note for one-time transfers', async () => {
```
Comment (jeff-clay) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx]: [Suggested tests] This test covers the 'user cleared the prefilled note' path. Consider adding a sibling test for 'user typed whitespace only' (`userEvent.type(noteField, '   ')`) to explicitly lock in the `.trim().min(1)` branch.

File: src/components/HrTools/SavingsFundTransfer/TransfersPage/TransfersPage.tsx [hunk 5]
```diff
@@ -84,6 +85,7 @@ export const TransfersPage: React.FC<TransfersPageProps> = ({ title }) => {
 
   const { data: staffAccountData, error: staffAccountError } =
     useStaffAccountQuery();
+  const { data: userData } = useGetUserQuery();
```
Comment (jeff-clay) [inline src/components/HrTools/SavingsFundTransfer/TransfersPage/TransfersPage.tsx]: [Mentoring] Calling `useGetUserQuery()` here is free in practice — `UserPreferenceProvider` calls the same operation at app root, so Apollo's normalized cache serves this read from memory. No need to thread `lastName` down from a higher ancestor.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx (hunk 4 above)
Comment (jeff-clay) [reply src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx]: https://github.com/CruGlobal/mpdx-react/pull/1772/commits/417b0af642d7eb5bb6cebae3add0e815b46a051d

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx (hunk 1 above)
Comment (jeff-clay) [reply src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx]: https://github.com/CruGlobal/mpdx-react/pull/1772/commits/417b0af642d7eb5bb6cebae3add0e815b46a051d

Comment (jeff-clay) [reply src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx]: From Claude: Implementing enableReinitialize on Formik would also re-run initialValues on any prop change, which could clobber user input in some edge cases — the cure is potentially worse than the rare race.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 6]
```diff
@@ -135,18 +135,28 @@ const transferSchema = (locale: string) =>
       .number()
       .required(i18n.t('Amount is required'))
       .min(0.01, i18n.t('Amount must be at least $0.01')),
-    note: yup.string().nullable(),
+    note: yup.string().when('schedule', {
+      is: ScheduleEnum.OneTime,
+      then: (schema) =>
+        schema
+          .trim()
+          .min(1, i18n.t('Note is required'))
```
Comment (dr-bizz) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx]: WThe min is redundant as you are trimming it and it is a required field. Unless you need more than 1 character on the backend.

File: src/components/HrTools/SavingsFundTransfer/TransfersPage/TransfersPage.tsx (hunk 5 above)
Comment (dr-bizz) [reply src/components/HrTools/SavingsFundTransfer/TransfersPage/TransfersPage.tsx]: Lets move this hook to the Model. There is no reason to do it here unless we are using `userData` on the component too.

It is best to just add it to the model, so it runs once it is open rather then it running when the page is loaded.

**Review by dr-bizz — APPROVED** [review]
This is looking good, but you don't need to add the `min` onto the `yup` logic as required does that already.

File: src/components/HrTools/SavingsFundTransfer/TransfersPage/TransfersPage.tsx (hunk 5 above)
Comment (jeff-clay) [reply src/components/HrTools/SavingsFundTransfer/TransfersPage/TransfersPage.tsx]: https://github.com/CruGlobal/mpdx-react/pull/1772/commits/6f5c546d9e52bc69e670b348d272c11ec0a8f48b

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx (hunk 6 above)
Comment (jeff-clay) [reply src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx]: https://github.com/CruGlobal/mpdx-react/pull/1772/commits/6f5c546d9e52bc69e670b348d272c11ec0a8f48b

**Comment by jeff-clay** [comment]
@dr-bizz Changes made. I know you approved already, but not sure if you want to do a review of changes. Not sure of protocol.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx [hunk 7]
```diff
@@ -68,6 +74,27 @@ interface ComponentsProps {
...
+        staffAccountId: null,
+        primaryDesignation: null,
+        userType: UserTypeEnum.UsStaff,
+        preferences: null,
+      },
+    },
+  });
+  return cache;
+};
+
```
Comment (dr-bizz) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx]: There is a different way to do this. Let me type it up

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx [hunk 8]
```diff
@@ -65,11 +71,34 @@ const transferDefaultData: TransferModalData['transfer'] = {
 interface ComponentsProps {
   transfer?: TransferModalData['transfer'];
   type?: TransferTypeEnum;
+  lastName?: string;
 }
 
+const buildCacheWithUser = (lastName: string) => {
```
Comment (dr-bizz) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx]: We don't need to build the cache. We can do the same as we've done here: src/components/DonationTable/DonationTable.test.tsx

We create the mocks on the GqlMockedProvider, which will mock the GQL queries

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 9]
```diff
@@ -135,7 +136,11 @@ const transferSchema = (locale: string) =>
       .number()
       .required(i18n.t('Amount is required'))
       .min(0.01, i18n.t('Amount must be at least $0.01')),
-    note: yup.string().nullable(),
+    note: yup.string().when('schedule', {
+      is: ScheduleEnum.OneTime,
```
Comment (dr-bizz) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx:140]: Are we only adding the required note if it is a one time payment?

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx [hunk 10]
```diff
@@ -80,6 +109,7 @@ const Components = ({
             createTransfer: CreateTransferMutation;
             updateRecurringTransfer: UpdateRecurringTransferMutation;
           }>
+            cache={buildCacheWithUser(lastName)}
```
Comment (dr-bizz) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx]: We can mock the response of the GQL query here so we don't need to store it in the cache. (I don't know if this will work out the box, as I haven't tested it, you might need to edit the GetUser)

We only need to enter the values we want to mock as we are randomly generating everything else.
```suggestion
            mocks={{
              GetUser {
                user {
                  id
                  firstName
                  lastName
                }
              }
            }}
```

**Review by dr-bizz — CHANGES_REQUESTED** [review]
Thank you for fixing that issue. We have a few more issues that have arisen.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx (hunk 9 above)
Comment (jeff-clay) [reply src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx:140]: Yeah, a few reasons:
- The note field currently only renders for one time transfers.
- The recurring GraphQL mutation doesn't accept a description
- SAA's recurring transfer synthesizes its own description server-side from the fund/account name.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx (hunk 10 above)
Comment (jeff-clay) [reply src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx]: https://github.com/CruGlobal/mpdx-react/pull/1772/commits/fcda99171e02ee31e4351fd20173f391ea72806c

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx (hunk 9 above)
Comment (jeff-clay) [reply src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx:140]: https://github.com/CruGlobal/mpdx-react/pull/1772/commits/fcda99171e02ee31e4351fd20173f391ea72806c

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx (hunk 10 above)
Comment (jeff-clay) [reply src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx]: Tried your `mocks` suggestion. It works for the mocks setup, but with `useGetUserQuery` inside the modal the data resolves async, and Formik captures `initialValues` at mount, so the default note ends up empty in that brief window.

Added a render-guard (`if (userLoading) return null`) as the smallest fix. In production it never fires (cache is warmed at app boot by `UserPreferenceProvider`). In tests it lets the mock resolve before assertions run, and let me drop the cache builder entirely.

If you'd prefer a different shape (e.g. compute the default via `useEffect` + `setFieldValue`, or not handle the loading case in the modal at all) happy to change. Just let me know.

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 11]
```diff
@@ -169,6 +174,19 @@ export const TransferModal: React.FC<TransferModalProps> = ({
   const type = data.type || TransferTypeEnum.New;
   const isNew = type === TransferTypeEnum.New;
 
+  const { data: userData, loading: userLoading } = useGetUserQuery();
+
+  if (userLoading) {
+    return null;
+  }
```
Comment (dr-bizz) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx:181]: We don't want to do this, as it can refetch data, and then we would no longer render the component.

See the next comment

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 12]
```diff
@@ -169,6 +174,19 @@ export const TransferModal: React.FC<TransferModalProps> = ({
...
+
+  const trimmedLastName = userData?.user?.lastName?.trim();
+  const defaultNote = trimmedLastName
+    ? t('{{lastName}} Savings Fund Transfer in MPDX', {
+        lastName: trimmedLastName,
+      })
+    : '';
+
   const title =
     type === TransferTypeEnum.New
```
Comment (dr-bizz) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.tsx:191]: We can use useMemo to reinitialise the variable `defaultNote` upon `userData` changing.
```suggestion
  const { data: userData } = useGetUserQuery();

 const defaultNote = useMemo(() => {
    const trimmedLastName = userData?.user?.lastName?.trim();
    return trimmedLastName
      ? t('{{lastName}} Savings Fund Transfer in MPDX', {
          lastName: trimmedLastName,
        })
      : '';
  }, [userData, t]);

  const title =
    type === TransferTypeEnum.New
```

File: src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx [hunk 13]
```diff
@@ -99,9 +108,15 @@ const Components = ({
   </SnackbarProvider>
 );
 
+const renderModal = async (props: ComponentsProps = {}) => {
+  const result = render(<Components {...props} />);
+  await result.findByText(/Fund Transfer/);
+  return result;
+};
```
Comment (dr-bizz) [inline src/components/HrTools/SavingsFundTransfer/TransferModal/TransferModal.test.tsx:115]: With now we're using useMemo, I'm unsure if you need this.

**Review by dr-bizz — CHANGES_REQUESTED** [review]
Thank you for making those edits. MPDX React is a large app, which means we need to keep it consistent throughout.

I noticed some more issues and have added my suggestions for you to implement.

The loading from GraphQL gets triggered every time it loads, both during the initial load and subsequent sequential loads.

I believe we can resolve this issue with a useMemo, which updates the default value when userData changes.

**Comment by BradleySchlather** [comment]
A new PR was opened for this ticket. That [PR](https://github.com/CruGlobal/mpdx-react/pull/1825) was reviewed and merged to main.
